### DISCUSSION TITLE
How can I define the range of negative sampling in the LinkNeighborLoader?

### DISCUSSION THREAD

**Discussion by songsong0425** [pyg-team/pytorch_geometric#9142] [Q&A] [answered]
Hi. Although I think that there would be a duplicated question, I have a question about the negative sampling in the `LinkNeighborLoader`.
Is there a way to select the negative samples from the pre-defined range of nodes?
For example, when I set the loader as below, it contains the unnecessary negative `edge_label_index`.
```python
train_loader = LinkNeighborLoader(kf_train_data, edge_label_index=kf_train_data.edge_label_index, edge_label=kf_train_data.edge_label,
                                      batch_size=32, shuffle=True, neg_sampling_ratio=1.0, num_neighbors=[2,2], disjoint=True, neg_sampling="binary")
```
I guess that the unknown edges in the `edge_index` of the training dataset were selected randomly as the negative samples in the batch.
(But it's a little weird because the documentation mentioned the `edge_label_index` and `edge_lablel` in the `neg_sampling` )

> neg_sampling ([NegativeSampling](https://pytorch-geometric.readthedocs.io/en/latest/modules/sampler.html#torch_geometric.sampler.NegativeSampling), optional) – The negative sampling configuration. For negative sampling mode "binary", samples can be accessed via the attributes edge_label_index and edge_label in the respective edge type of the returned mini-batch. In case edge_label does not exist, it will be automatically created and represents a binary classification task

When I have a node list and want to sample the random negative edges avoiding the duplicated edges with true positive `edge_label_index`, how can I extract random negative samples?
```python
print(node_list)
print(len(node_list))
# array([     0,      1,      3, ..., 101089, 101090, 101091])
# 19392
```

Currently, I created random edges from the node list before entering the loader, but since I have modified the dataset it is difficult to sample new negative samples during the epoch.
```python
random_neg_edges = torch.stack([torch.tensor(random.sample(list(node_list), len(GraphList.edge_label))), 
                                    torch.tensor(random.sample(list(node_list), len(GraphList.edge_label)))], 0)
GraphList.edge_index = torch.cat([GraphList.edge_index, random_neg_edges], dim=1)
GraphList.edge_label = torch.cat([GraphList.edge_label, torch.zeros(len(GraphList.edge_label)).int()])
```

**Comment by rusty1s** [comment]
Sorry, I don't understand the question yet. Why is `edge_label_index` useless? `edge_label_index` refers to the set of edges we want to train against, and they are extended by negative edges as well. Positive and negative edges can then be discriminated via `edge_label` (same number of elements as positive/negative edges).

If you want to sample on your own, you can also take a look at `utils.negative_sampling`.

**Reply by songsong0425** [reply]
I'm sorry for the poor question. I hope the below figure helps your understanding.
(If it is wrong, please let me know)
<img width="537" alt="sampling" src="https://github.com/pyg-team/pytorch_geometric/assets/80442548/26b3b12e-aae4-4a00-bf00-3d45dc134e55">
In short, I only want to sample the negative edges from the unknown links between blue nodes.
But when I set the `neg_sampling_ratio=1.0`, the `loader` automatically sampled edges from the `edge_index` containing black nodes. (I need to calculate the probability and evaluation metrics only between blue nodes)
Can I prevent this phenomenon?

**Reply by rusty1s** [reply, marked as answer]
We currently lack a good interface to customize negative sampling. What you can try is passing a custom `NegativeSampling` class to your loader, which modifies `sample` to only sample blue nodes.
